libs/cache/file_cache.go

```diff
@@ -0,0 +1,309 @@
+	userCacheDir, err := os.UserCacheDir()
+	if err != nil {
+		return "", fmt.Errorf("failed to get user cache directory: %w", err)
+	}
+	return filepath.Join(userCacheDir, "databricks"), nil
+}
+
+// NewFileCache creates a new file-based cache using UserCacheDir() + "databricks" + cached component name.
+func NewFileCache[T any](component string, expiryMinutes int, metrics *bundle.Metrics) (*FileCache[T], error) {
+	cacheBaseDir, err := getCacheBaseDir()
+	if err != nil {
+		return nil, err
+	}
+
+	baseDir := filepath.Join(cacheBaseDir, component)
+	fc, err := newFileCacheWithBaseDir[T](baseDir, expiryMinutes)
+	if err != nil {
+		return nil, err
+	}
+	fc.metrics = metrics
+	return fc, nil
+}
+
+// cacheEntry represents the structure of a cached item on disk.
+type cacheEntry struct {
+	Data      json.RawMessage `json:"data"`
+	Expiry    time.Time       `json:"expiry"`
+	Timestamp time.Time       `json:"timestamp,omitempty"` // For backward compatibility
+}
+
+func (fc *FileCache[T]) addTelemetryMetric(key string) {
+	if fc.metrics != nil {
+		fc.metrics.SetBoolValue(key, true)
+	}
+}
+
+// GetOrCompute retrieves cached content or computes it using the provided function.
+func (fc *FileCache[T]) GetOrCompute(ctx context.Context, fingerprint any, compute func(ctx context.Context) (T, error)) (T, error) {
+	var zero T
+	isCacheHit := false
+
+	// Convert fingerprint to deterministic string
+	fingerprintHash, err := fingerprintToHash(fingerprint)
+	log.Debugf(ctx, "[Local Cache] using fingerprint with hash: %s\n", fingerprintHash)
+
+	fc.addTelemetryMetric("local.cache.attempt")
+
+	if err != nil {
+		log.Debugf(ctx, "[Local Cache] cache miss: non-compliant fingerprint\n")
+		return zero, fmt.Errorf("failed to convert fingerprint to string: %w", err)
+	}
+
+	cacheKey := fc.getCacheKey(fingerprintHash)
+	log.Debugf(ctx, "[Local Cache] using cache key: %s\n", cacheKey)
+
+	cachePath := fc.getCachePath(cacheKey)
+	log.Debugf(ctx, "[Local Cache] using cache path: %s\n", cachePath)
+
+	// Check in-memory cache first
+	fc.mu.RLock()
+	if _, found := fc.memCache[cacheKey]; found {
+		fc.mu.RUnlock()
+		log.Debugf(ctx, "[Local Cache] cache hit: in-memory\n")
+		fc.addTelemetryMetric("local.cache.hit")
+		isCacheHit = true
+		// return data, nil	// cache layer is currently no-op
+	} else {
+		fc.mu.RUnlock()
+	}
+
+	// Try to read from disk cache
+	if data, found := fc.readFromCache(cachePath); found {
+		// Store in memory cache for faster future access
+		fc.mu.Lock()
+		fc.memCache[cacheKey] = data
+		fc.mu.Unlock()
+		log.Debugf(ctx, "[Local Cache] cache hit: disk-read\n")
+		fc.addTelemetryMetric("local.cache.hit")
+		isCacheHit = true
+		// return data, nil // cache layer is currently no-op
+	}
+
+	// Check if there's a pending write for this key
+	fc.mu.Lock()
+	if pendingCh, exists := fc.pending[cacheKey]; exists {
+		fc.mu.Unlock()
+		// Wait for pending write to complete
+		select {
+		case <-pendingCh:
+			// Try reading from memory cache again
```

Comment:
It seems like there could be a race condition between unlocking (fc.mu.Unlock()) and re-acquiring the read lock (fc.mu.RLock()) when another goroutine could modify memCache